src/main/java/servlets/module/lesson/XxeLesson.java

```diff
@@ -0,0 +1,174 @@
+ * <p>
+ * You should have received a copy of the GNU General Public License
+ * along with the Security Shepherd project.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * @author ismisepaul
+ */
+public class XxeLesson
+        extends HttpServlet {
+    private static final long serialVersionUID = 1L;
+    private static Logger log = Logger.getLogger(XxeLesson.class);
+    private static final String LEVEL_NAME = "XXE Lesson";
+    private static final String LEVEL_ID = "df2ac757cc135dcb8ce5ea01f677c74f04b446d6";
+    private static final String LEVEL_HASH = "57dda1bf9a2ca1c34e04f815491ef40836d9b710179cd19754ec5b3c31f27d1a";
+
+    public void doPost(HttpServletRequest request, HttpServletResponse response)
+            throws ServletException, IOException {
+        boolean moduleOpen = false;
+
+
+        //Setting IpAddress To Log and taking header for original IP if forwarded from proxy
+        ShepherdLogManager.setRequestIp(request.getRemoteAddr(), request.getHeader("X-Forwarded-For"));
+        log.debug(LEVEL_NAME + " Servlet Accessed");
+        PrintWriter out = response.getWriter();
+        out.print(getServletInfo());
+
+        //Translation Stuff
+        Locale locale = new Locale(Validate.validateLanguage(request.getSession()));
+        ResourceBundle errors = ResourceBundle.getBundle("i18n.servlets.errors", locale);
+        ResourceBundle bundle = ResourceBundle.getBundle("i18n.servlets.lessons.xxe", locale);
+
+        try {
+            HttpSession ses = request.getSession(true);
+            if (Validate.validateSession(ses))
+            {
+                if (Getter.isModuleOpen(getServletContext().getRealPath(""), LEVEL_ID))
+                {
+                    ShepherdLogManager.setRequestIp(request.getRemoteAddr(), request.getHeader("X-Forwarded-For"),
+                            ses.getAttribute("userName").toString());
+                    log.debug(LEVEL_NAME + " accessed by: " + ses.getAttribute("userName").toString());
+                    Cookie tokenCookie = Validate.getToken(request.getCookies());
+                    Object tokenHeader = request.getHeader("csrfToken").toString();
+
+                    if (Validate.validateTokens(tokenCookie, tokenHeader)) {
+                        InputStream xml = request.getInputStream();
+                        String emailAddr = readXml(xml);
+                        log.debug("Email Addr: " + emailAddr);
+
+                        String htmlOutput = new String();
+
+                        if (emailAddr == null) {
+                            htmlOutput += "<p>" + bundle.getString("response.blank.email") + "</p>";
+                            out.write(htmlOutput + emailAddr);
+                        } else if (Validate.isValidEmailAddress(emailAddr)) {
+                            log.debug("User Submitted - " + emailAddr);
+
+                            htmlOutput += "<p>" + bundle.getString("response.success.reset") + ": " + emailAddr
```

Comment:
Shouldn't emailAddr be encoded for HTML before added to the htmlOutput string?

Reply:
The encoding is being done in the readXml() function in that class. The return function encodes it.

https://github.com/OWASP/SecurityShepherd/blob/2cac9ef1673aa8716253a645be66ef3db4145a58/src/main/java/servlets/module/lesson/XxeLesson.java#L135